Exporting gradients of integrals to python

ashutoshvt opened the pull request:
## Description
This PR aims to export the first and second derivatives of one and two electron integrals in both AO and MO basis to python so that they can be used in psi4numpy for educational purposes. While calculating first derivatives, one gets a vector of matrices as the output, one for each cartesian component of the atom number provided in the input. Similarly, for hessians, one gets a total of 9 matrices for a given pair of atoms. 
## Todos
Notable points that this PR has either accomplished or will accomplish.
* **Developer Interest**
  - [X] add first derivatives 
  - [X] add second derivatives
  - [X] optimization and cleanup
  - [X] add corresponding test cases in psi4numpy tests folder

## Status
- [X] Ready to go

dgasmith commented:
@ashutoshvt If you are delayed with other things it might be wise for us to just patch up first derivates and get that in. We can work on the rest in other PR's.

ashutoshvt commented:
@dgasmith Sure. I  will let you know soon (hopefully this Tuesday) after resolving some conflicts.

ashutoshvt commented:
@dgasmith Just wanted to let you know that the code works now but is a bit slow and I am trying to make it faster.

dgasmith commented:
@ashutoshvt I believe this will only be used in experimental code as you likely should contract the integrals on the fly for efficient code. Nothing really stands out at me as coded in a way that would take a lot of extra time, what kind of performance improvements are you looking for?

ashutoshvt commented:
I know this is meant for educational purposes but for a basis size of 40, computing AO hessian of TEIs of a pair of atoms took around 3.5 minutes. To be honest, I don't exactly know why its so slow at this moment. But I want to explore a bit more.

ashutoshvt commented:
@dgasmith I think this is ready for review now.

dgasmith commented on psi4/src/psi4/libmints/mintshelper.cc:
```diff
@@ -793,13 +797,13 @@ SharedMatrix MintsHelper::ao_helper(const std::string &label, std::shared_ptr<Tw
     int pfxn = basisset_->shell(P).nfunction();
     int qfxn = basisset_->shell(Q).nfunction();
-    auto I = std::make_shared<Matrix>(label, mfxn * nfxn, pfxn * qfxn);
+    SharedMatrix I(new Matrix(label, mfxn * nfxn, pfxn * qfxn));
```
It looks like you reversed a few of the `make_shared` changes. Can you look at patching that up?

ashutoshvt replied:
Indeed, the rebasing led to these. I will patch them up.

dgasmith commented on psi4/src/psi4/libmints/mintshelper.cc:
```diff
@@ -822,89 +826,111 @@ SharedMatrix MintsHelper::ao_shell_getter(const std::string &label, std::shared_
+{
+    std::shared_ptr <BasisSet> bs1 = ints->basis1();
+    std::shared_ptr <BasisSet> bs2 = ints->basis2();
+    std::shared_ptr <BasisSet> bs3 = ints->basis3();
```
Yea, it looks like when you rebased you flipped a lot of the clang-format changes. Could you re-apply clang format to MintsHelper? @robertodr can probably help you locally.

ashutoshvt replied:
Sure. I will try clang formatting MintsHelper.

dgasmith commented on psi4/src/psi4/libmints/mintshelper.cc:
```diff
@@ -1510,22 +1557,25 @@ std::vector<SharedMatrix> MintsHelper::ao_nabla() {
 
-    if ((C1->nirrep() + C2->nirrep() + C3->nirrep() + C4->nirrep()) > 4) {
-        throw PSIEXCEPTION("MO Transform: Incoming orbitals must be C1 symmetry.");
-    }
```
Any reason to remove this? I still think its needed?

ashutoshvt replied:
Most of these changes are related to me not resolving the merge conflicts properly while rebasing.

dgasmith commented on psi4/src/psi4/libmints/mintshelper.cc:
```diff
@@ -1996,4 +2046,1146 @@ void MintsHelper::play()
+
+        std::vector<SharedMatrix> grad;
+        for (int p=0; p<3; p++){
+            sprintf(lbl1, "%s%d_%s", lbl, atom, cartcomp[p]);
```
Can you just use a `std::string` or `std:stringstream` here?

ashutoshvt replied:
Done!

dgasmith commented on psi4/src/psi4/libmints/mintshelper.cc:
```diff
@@ -1996,4 +2046,1146 @@ void MintsHelper::play()
+        cartcomp = (char **) malloc (3 * sizeof(char *));
+        cartcomp[0] = strdup("X");
+        cartcomp[1] = strdup("Y");
+        cartcomp[2] = strdup("Z");
```
Great place to use `std:vector<std::string>`.

ashutoshvt replied:
Done!

dgasmith commented on psi4/src/psi4/libmints/mintshelper.cc:
```diff
@@ -1996,4 +2046,1146 @@ void MintsHelper::play()
+            grad.push_back(SharedMatrix(new Matrix(lbl1, nbf1, nbf2)));
+            }
+
+        const double* buffer = GInt->buffer();
```
If you dont use the std library make sure to pop `cartcomp`.

ashutoshvt replied:
I am using the std library.

dgasmith commented on psi4/src/psi4/libmints/mintshelper.cc:
```diff
@@ -1996,4 +2046,1146 @@ void MintsHelper::play()
+                   vec_pairs.push_back(std::make_pair(atom,aQ)); 
+
+
+                   for (std::vector<std::pair<int, int> >::iterator it = vec_pairs.begin() ; it != vec_pairs.end(); ++it){
```
Could be nice to use a ranged based for loop here.

ashutoshvt replied:
I need the position of the elements of the container here. So a range based for loop may not be ideal here.

dgasmith commented on tests/psi4numpy/rhf_gradient/input.py:
```diff
@@ -0,0 +1,137 @@
+                if key == "S":
+                    Gradient[key][atom][p] = -2.0 * np.einsum("ii,ii->", F[:occ,:occ], deriv1_np[map_key][:occ,:occ])
+                else:
+                    Gradient[key][atom][p] = 2.0 * np.einsum("ii->", deriv1_np[map_key][:occ,:occ])
```
This looks awesome! Glad these are in Psi4Numpy format, will you drop them into the main Psi4NumPy repo as well?

ashutoshvt replied:
Sure, i will create a PR in Psi4Numpy soon.

dgasmith commented on tests/psi4numpy/rhf_gradient/input.py:
```diff
@@ -0,0 +1,137 @@
+             [-0.000000000000, -0.086300100260,  0.048720720189],
+             [-0.000000000000,  0.086300100260,  0.048720720189]
+       ])
+G_python = rhf_gradient(mol, options)
```
Something that could be nice here as well is to compare to the new gradient tech in MintsHelper `kinetic_grad`, `potential_grad`, etc. These terms contract the density direction with the gradient to build a `atom x 3` matrix directly which should be identical to the one-electron components that you form here.

ashutoshvt replied:
Done!

ashutoshvt commented:
@dgasmith I think have addressed all the concerns now. I just realized that I should have clang-formatted the files in a separate commit. Sorry about that!

dgasmith commented on tests/psi4numpy/rhf_gradient/input.py:
```diff
@@ -0,0 +1,163 @@
+    F = H + 2.0 * np.einsum('pmqm->pq', MO[:, :occ, :, :occ])
+    F -= np.einsum('pmmq->pq', MO[:, :occ, :occ, :])
+    natoms = mol.natom()
+    cart = {}; cart[0] = '_X'; cart[1] = '_Y'; cart[2] = '_Z'
```
It looks like you are building a set here? `{'_x', '_y', '_z'}` will do the same thing.

JonathonMisiewicz replied:
{} creates a dict, not a set.

On that subject, might I suggest using a tuple here? When all you need is to access elements based on an index, a tuple will suffice and is more Pythonic.

dgasmith replied:
Ah, yes. Too early in the morning. Either way, best to do it as an explicit build.

ashutoshvt replied:
I think a list is the best way to go in my case as set doesn't support indexing and a tuple might be unnecessary for this simple example.

dgasmith commented on tests/psi4numpy/rhf_gradient/input.py:
```diff
@@ -0,0 +1,163 @@
+# Test OEI gradients with that of PSI4
+psi4.compare_matrices(PSI4_Grad["S"], G_python_S_mat, 10, "OVERLAP_GRADIENT_TEST")   # TEST 
+psi4.compare_matrices(PSI4_Grad["T"], G_python_T_mat, 10, "KINETIC_GRADIENT_TEST")   # TEST
+psi4.compare_matrices(PSI4_Grad["V"], G_python_V_mat, 10, "POTENTIAL_GRADIENT_TEST") # TEST
```
Cool! Good to check this both ways.

dgasmith commented on psi4/src/export_mints.cc:
```diff
@@ -68,7 +68,6 @@
 #include "psi4/libmints/dipole.h"
```
I dont know if you clang-format is picking up our limited style sheet as there shouldn't be this many changes in this file. These options are:
```
BasedOnStyle: Google
IndentWidth: 4
ColumnLimit: 120
SortIncludes: false
```
and can be found in `psi4/.clang-format`. It looks like you are truncating at 90 line-width, not 120.

ashutoshvt replied:
I am using the standard command: `clang-format -style=file -i export_mints.cc`. This should automatically use the `psi4/.clang-format`file for formatting.

dgasmith commented:
@psi4/maintainers Can we get a review on this, I think its time to get this in. 

I believe @ashutoshvt Will edit the psi4numpy tests, but thats it.

andysim commented on tests/psi4numpy/rhf_hessian/input.py:
```diff
@@ -0,0 +1,284 @@
+F = H + 2.0 * np.einsum('pmqm->pq', MO[:, :occ, :, :occ])
+F -= np.einsum('pmmq->pq', MO[:, :occ, :occ, :])
+natoms = mol.natom()
+cart = {}; cart[0] = '_X'; cart[1] = '_Y'; cart[2] = '_Z'
```
This is only used as an array, not a dict, as far as I can tell.  Maybe `cart = ['_X', '_Y', '_Z']` is a little more Pythonic in that case.

ashutoshvt commented:
Thanks @andysim. Changed the dict to list.

loriab commented on tests/psi4numpy/rhf_hessian/CMakeLists.txt:
```diff
@@ -0,0 +1,3 @@
+include(TestingMacros)
+
+add_regression_test(psi4numpy-rhf_hessian "psi;quicktests;smoketests;psi4numpy")
```
Don't need both grad & hess in smoketests, as they're unlikely to break independently.

ashutoshvt replied:
Done!

loriab commented on tests/psi4numpy/rhf_gradient/CMakeLists.txt:
```diff
@@ -0,0 +1,3 @@
+include(TestingMacros)
+
+add_regression_test(psi4numpy-rhf_gradient "psi;quicktests;smoketests;psi4numpy")
```
tests use dash rather than underscore in their names, so that'll need altering in several places.

ashutoshvt replied:
Done.

loriab commented on tests/psi4numpy/rhf_gradient/input.py:
```diff
@@ -0,0 +1,163 @@
+G_python_V_mat = psi4.core.Matrix.from_array(Gradient["V"])
+
+# Test OEI gradients with that of PSI4
+psi4.compare_matrices(PSI4_Grad["S"], G_python_S_mat, 10, "OVERLAP_GRADIENT_TEST")   # TEST 
```
actually has to be `#TEST` w/o space

ashutoshvt replied:
Done.

loriab commented on tests/psi4numpy/rhf_gradient/input.py:
```diff
@@ -0,0 +1,163 @@
+Reference: "Derivative studies in hartree-fock and møller-plesset theories",
+J. A. Pople, R. Krishnan, H. B. Schlegel and J. S. Binkley
+DOI: 10.1002/qua.560160825
+"""
```
I'm not sure what about this file (and hess) is offending py27, but according to travis, looks like something is. maybe build a 27 locally to find out why. or maybe that's what [this](https://github.com/ashutoshvt/psi4/blob/b3c4612f9ef0657d8d6e9e4cbdbcf99136bebed4/tests/libxc/input.dat#L135) is for.

ashutoshvt replied:
Fixed it. Python 2 was not able to handle the non-ascii character ø in the reference.

ashutoshvt commented on tests/psi4numpy/rhf_hessian/input.py:
```diff
@@ -0,0 +1,284 @@
+F = H + 2.0 * np.einsum('pmqm->pq', MO[:, :occ, :, :occ])
+F -= np.einsum('pmmq->pq', MO[:, :occ, :occ, :])
+natoms = mol.natom()
+cart = {}; cart[0] = '_X'; cart[1] = '_Y'; cart[2] = '_Z'
```
I already took care of this.

ashutoshvt commented on psi4/src/psi4/libmints/mintshelper.cc:
```diff
@@ -51,6 +52,10 @@
+#include <map>
+#include <unordered_map>
+#include <list>    
+#include <utility>
```
Didn't understand this point. Could you please elaborate?

loriab replied:
Just that order should be alphabetical

```
#include <cstdlib>
#include <cstdio>
#include <cmath>
#include <iostream>
#include <list>    
#include <map>
#include <sstream>
#include <unordered_map>
#include <utility>
#include <vector>
```

ashutoshvt replied:
OK. Thanks! Fixed it now

loriab commented on tests/psi4numpy/CMakeLists.txt:
```diff
@@ -1,4 +1,6 @@
 add_subdirectory(cphf)
 add_subdirectory(dfmp2)
 add_subdirectory(rhf)
+add_subdirectory(rhf_gradient)
```
underscore to dash here, too, I think

ashutoshvt replied:
Sorry! Too late to be working, it seems :)
